### PR TITLE
Override proxy minion opts with pillar data

### PR DESCRIPTION
**Pull request by mirceaulinic**
### What does this PR do?

Overrides proxy minion opts with their pillar data.
Potential problem: it copies also everything under the `proxy` key (including sensitive data etc.) into the opts. Do you think this may be a problem?
To solve #39775 speficially, it would be enough just to transfer the values of the `mine_*` fields into the opts object. But I was thinking that there may be other interesting cases to solve by copying everything.
I am looking forward to hearing your thoughts on this. Thanks!

### What issues does this PR fix or reference?

#39775

### Previous Behavior

`mine_interval` ignored in the proxy pillar

### New Behavior

`mine_interval` and other options have the values per minion

### CODE REVIEW

**Comment by mirceaulinic** [comment]
Ping @cro

File: salt/minion.py
Comment (mirceaulinic) [inline salt/minion.py]: Os should we apply a proper merge and update similar to `config.merge`?

i.e. having

```yaml
my_option:
  my_suboption1: val1
```
in the opts, and:

```yaml
my_option:
  my_suboption2: val2
```

With these changes, will make the opts:

```yaml
my_option:
  my_suboption2: val2
```

Do we want something like below?

```yaml
my_option:
  my_suboption1: val1
  my_suboption2: val2
```

**Comment by cachedout** [comment]
This needs to be reviewed by @cro before it can be merged.

**Comment by mirceaulinic** [comment]
Till @cro replies: I played a bit with this, I wasn't able to identify any potential risk. If we really want to, we can pop the `proxy` key, but if someone can do `opts['proxy']['password']`, can equally do `opts['pillar']['proxy']['password']`.

I was thinking that we could enhance the merging with few options:

- `proxy_opts_pillar_merge`, defaulting to True: merge the pillar in the opts.
- `proxy_opts_pillar_deep_merge`, defaulting to False: deep merge.
- `proxy_opts_pillar_merge_strategy`, defaulting to `smart`, active when `proxy_opts_pillar_deep_merge` is enabled.
- `proxy_opts_pillar_merge_lists`, defaulting to `False`: similar to the `pillar_merge_lists` option on the master.

I know there are quite a few options introduced for a simple improvement, but I think it's safer like that.
At the same time, I am wondering (as I don't have much experience with regular minions): how is the pillar merged in the minion opts? Reading the code I didn't get it...

If my thoughts above make sense, please let me know and I will go ahead to implement and document. Thanks!

**Comment by whiteinge** [comment]
Still catching up on this issue, but brief thoughts: I'm not sure there's a precedent for merging Pillar into opts. I think Pillar overrides fall more into "run time" overrides and continually-running things like mine and scheduler take a more dynamic path to getting updated when Pillar is refreshed.

**Comment by mirceaulinic** [comment]
Thanks for clarifying this @whiteinge.
I was a bit confused by https://docs.saltstack.com/en/latest/topics/mine/:

> To enable the Salt Mine the mine_functions option needs to be applied to a Minion. This option can be applied via the Minion's configuration file, or the Minion's Pillar.

Currently, setting `mine_functions` or `mine_interval` under the pillar of a proxy minion, the option is ignored.
Therefore I would have some questions to sort this out:

1. Do we want other options to be configured from the pillar, or just `mine_*`?
2. If we want others, merging should be by default, or enabled through an option?
3. if (2), do we want options for strategy and lists?

Thanks!

**Comment by whiteinge** [comment]
I'd need to double-check this claim but I think non-proxy mine/scheduler use `config.get` to look for config values in either minion conf, master conf, and pillar. I'm not sure how they are triggered when pillar is refreshed. I'll check in with the core team and reply back.

**Comment by mirceaulinic** [comment]
Thanks, @whiteinge.

Either way, I have identified cases when some proxy minions require different functions to be cached in the mine. Typically the proxy minions run on the same machine, regardless of their proxy type, hence they all are using the same proxy configuration file, `/etc/salt/proxy`.
But different proxies have a different set of functions they can execute, hence the mine caching needs to be different depending on what proxy type they use. In that case, we need more granular control (at least for mines, but perhaps there could be other edge cases like that).

Does this make any sense? If I wasn't able to expose the idea properly, please let me know.

**Comment by mirceaulinic** [comment]
Hi - just checking in here if my thoughts above make sense?

I would take @cro's lack of reply as a silent acceptance.
But please let me know if this needs more work (specifically merge only the `mine_*` options, although there may be others that may need more granular control).

Thanks!

**Comment by rallytime** [comment]
I have switched the base branch on this pull request from `nitrogen` to `2017.7` in preparation for moving from a code-named release to a version number.

**Comment by rallytime** [comment]
Go Go Jenkins!

**Review by cro — CHANGES_REQUESTED** [review]
However this falls out I think it would be best if the proxy behavior did not deviate significantly from regular minions.

As this PR stands, though, it sidesteps the need for some customers to have proxy minions that get their configuration *only* from their /etc/salt/proxy file.

I see the value in the approach though, so we probably need to config gate this somehow with some options that will accommodate the different use cases.

**Comment by mirceaulinic** [comment]
Thanks for looking @cro. I have listed above the following four options:

- `proxy_opts_pillar_merge`, defaulting to True: merge the pillar in the opts.
- `proxy_opts_pillar_deep_merge`, defaulting to False: deep merge.
- `proxy_opts_pillar_merge_strategy`, defaulting to `smart`, active when `proxy_opts_pillar_deep_merge` is enabled.
- `proxy_opts_pillar_merge_lists`, defaulting to `False`: similar to the `pillar_merge_lists` option on the master.

Do they make sense? Thoughts?

**Comment by mirceaulinic** [comment]
Ping @cro

**Comment by cachedout** [comment]
@mirceaulinic Sounds like @cro is busy today. I chatted with him and he'll try to get to this one tomorrow.

**Comment by mirceaulinic** [comment]
Many thanks @cachedout

**Comment by mirceaulinic** [comment]
Ping @cro

**Comment by mirceaulinic** [comment]
As suggested in my comment above, I have added a couple of options regarding the merging of pillar data into opts: https://github.com/saltstack/salt/pull/41547/commits/5171b2a59ebf5fc1c881c5a7115eaa3c59c496cf
Although I have suggested`proxy_opts_pillar_merge` to default to True, I think it does make sense to default it to False.

Regarding the mines settings, it does make sense to merge the pillar settings into opts by default, however, this can be changed using the  `proxy_mines_pillar` option: https://github.com/saltstack/salt/pull/41547/commits/13f75330617ab1069fcdf53139d348da44152be6.

If these changes make sense, please let me know and I'll update the documentation.

**Comment by mirceaulinic** [comment]
Ping @cro @cachedout - any opinions?

**Comment by mirceaulinic** [comment]
Ping @cro

**Comment by mirceaulinic** [comment]
Hi @cro - do you think there would be any possibility for you to re-review this please? It's been more than two months already and this is definitely not encouraging me to try fixing this kind of bugs.
At the same time I understand that @cro doesn't work anymore on the open version (though the activity under the `salt-users` would suggest something different), so probably it would be indicated that someone else to review this PR and other proxy-related issues? I'd really appreciate some activity in this area.

In other train of thoughts, there's the following documentation for the pillar: https://docs.saltstack.com/en/latest/topics/pillar/#minion-config-in-pillar

> Minion configuration options can be set on pillars. Any option that you want to modify, should be in the first level of the pillars, in the same way you set the options in the config file.

I believe this should be enough argument to preserve the same behaviour for the proxy minions, including mine-specific options too. Would you disagree with that?

Thanks & looking forward for an answer!

**Review by cro — APPROVED** [review]

**Comment by mirceaulinic** [comment]
Hi @cachedout @cro I have fast forwarded this branch and documented the new opts. Thanks!

**Pull request merged**
